libraries/SITL/SIM_GPS.cpp

```diff
@@ -1014,6 +1017,311 @@ uint32_t GPS::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc
+
+    // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%201
+    enum class POS_FLAGS_1 : uint8_t {
+        NEW_POSITION = 1U << 0,
+        CLOCK_FIX_CALULATED = 1U << 1,
+        HORIZ_FROM_THIS_POS = 1U << 2,
+        HEIGHT_FROM_THIS_POS = 1U << 3,
+        RESERVED_4 = 1U << 4,
+        LEAST_SQ_POSITION = 1U << 5,
+        RESERVED_6 = 1U << 6,
+        POSITION_L1_PSEUDORANGES = 1U << 7
+    };
+    const uint8_t pos_flags_1 {
+        uint8_t(POS_FLAGS_1::NEW_POSITION) |
+        uint8_t(POS_FLAGS_1::CLOCK_FIX_CALULATED) |
+        uint8_t(POS_FLAGS_1::HORIZ_FROM_THIS_POS) |
+        uint8_t(POS_FLAGS_1::HEIGHT_FROM_THIS_POS) |
+        uint8_t(POS_FLAGS_1::RESERVED_4) |
+        uint8_t(POS_FLAGS_1::LEAST_SQ_POSITION) |
+        uint8_t(POS_FLAGS_1::POSITION_L1_PSEUDORANGES)
+    };
+
+    // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%202
+    enum class POS_FLAGS_2 : uint8_t {
+        DIFFERENTIAL_POS = 1U << 0,
+        DIFFERENTIAL_POS_PHASE_RTK = 1U << 1,
+        POSITION_METHOD_FIXED_PHASE = 1U << 2,
+        OMNISTAR_ACTIVE = 1U << 3,
+        DETERMINED_WITH_STATIC_CONSTRAINT = 1U << 4,
+        NETWORK_RTK = 1U << 5,
+        DITHERED_RTK = 1U << 6,
+        BEACON_DGNSS = 1U << 7,
+    };
+
+    // Simulate a GPS without RTK in SIM since there is no RTK SIM params.
+    // This means these flags are unset:
+    // NETWORK_RTK, DITHERED_RTK, BEACON_DGNSS
+    uint8_t pos_flags_2 {0};
+    if(d->have_lock) {
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::DIFFERENTIAL_POS);
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::DIFFERENTIAL_POS_PHASE_RTK);
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::POSITION_METHOD_FIXED_PHASE);
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::OMNISTAR_ACTIVE);
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::DETERMINED_WITH_STATIC_CONSTRAINT);
+    }
+
+    const auto gps_tow = gps_time();
+    struct PACKED gsof_pos_time {
+        const uint8_t OUTPUT_RECORD_TYPE { GSOF_POS_TIME_TYPE }; 
+        const uint8_t RECORD_LEN { GSOF_POS_TIME_LEN };
+        uint32_t time_week_ms;
+        uint16_t time_week;
+        uint8_t num_sats;
+        // TODO
+        // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%201
+        uint8_t pos_flags_1;
+        // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%202
+        uint8_t pos_flags_2;
+        // TODO increment every time GPS is "restarted", can we use boot count?
+        uint8_t initialized_num = bootcount; 
+    } pos_time;
+    static_assert(sizeof(gsof_pos_time) - (sizeof(gsof_pos_time::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos_time::RECORD_LEN)) == GSOF_POS_TIME_LEN);
+    pos_time.pos_flags_1 = pos_flags_1;
+    pos_time.pos_flags_2 = pos_flags_2;
+    pos_time.time_week_ms = htobe32(gps_tow.ms);
+    pos_time.time_week = htobe16(gps_tow.week);
+    pos_time.num_sats = d->have_lock ? _sitl->gps_numsats[instance] : 3;
```

Comment:
```suggestion
    const auto gps_tow = gps_time();
    const struct PACKED gsof_pos_time {
        const uint8_t OUTPUT_RECORD_TYPE;
        const uint8_t RECORD_LEN;
        uint32_t time_week_ms;
        uint16_t time_week;
        uint8_t num_sats;
        uint8_t pos_flags_1;
        uint8_t pos_flags_2;
        uint8_t initialized_num;
    } pos_time {
        GSOF_POS_TIME_TYPE,
        GSOF_POS_TIME_LEN,
        htobe32(gps_tow.ms),
        htobe16(gps_tow.week),
        d->have_lock ? _sitl->gps_numsats[instance] : uint8_t(3),
        pos_flags_1,
        pos_flags_2,
        bootcount
    };
    static_assert(sizeof(gsof_pos_time) - (sizeof(gsof_pos_time::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos_time::RECORD_LEN)) == GSOF_POS_TIME_LEN);
```

similarly for the other packets.

Reply:
I'll do this, but I'm leaving in the URL's to the documentation for each packet. They have been handy.

Reply:
Done!